File: Problems & Solutions/380. Insert Delete GetRandom O(1)/29 November 2022/Insert Delete GetRandom O(1).swift
final class RandomizedSet {
    var hashMap = [Int: Int]()
    var arr = [Int]()
    
    init() {
        
    }
    
    func insert(_ val: Int) -> Bool {
        if hashMap[val] != nil {
            return false
        }
        arr.append(val)
        hashMap[val] = arr.count - 1
        return true
    }
    
    func remove(_ val: Int) -> Bool {
        if hashMap[val] == nil {
            return false
        }
        let index = hashMap[val]!
        let lastVal = arr.last!
        arr.swapAt(index, arr.count - 1)
        hashMap[lastVal] = index
        arr.removeLast()
        hashMap.removeValue(forKey: val)
        return true
    }

    func getRandom() -> Int {
        return arr.randomElement()!
    }
}

File: Problems & Solutions/380. Insert Delete GetRandom O(1)/29 November 2022/Insert Delete GetRandom O(1).rb
class RandomizedSet
    def initialize()
        @hash = Hash.new
        @nums = Array.new
    end
    def insert(val)
        return false if @hash.key?(val)
        @nums << val
        @hash[val] = @nums.size - 1
        true
    end
    def remove(val)
        return false unless @hash.key?(val)
        @hash[@nums.last] = @hash[val] if val != @nums.last
        @nums[@hash[val]] = @nums.last
        @nums.delete_at(@nums.size - 1)
        @hash.delete(val)
        true
    end
    def get_random()
        @nums[rand(@nums.size)]
    end
end